### ISSUE TITLE
JSONAPIConverterFactory - compatibility with other response types

### ISSUE THREAD

**Issue by bradroid** [jasminb/jsonapi-converter#126]
I have existing retrofit instance built like this:

```
        retrofit = new Retrofit.Builder()
                .client(client)
                .baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
```

This instance of Retrofit is used to build API interface which contains a lot of legacy endpoints, not integrated with your library.

I would like to use the same `Retrofit` instance, modify it's creation to add JSONAPI converter  like this:

```
        retrofit = new Retrofit.Builder()
                .client(client)
                .baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addConverterFactory(jsonApiConverterFactory)
                .build();
```

Then, I would like to add new API endpoint definitions to existing retrofit interface, and I expect both old and new calls to work.

However, when I do this, the new calls (integrated with your library) work, but some of the old ones break.

By adding your JSONAPI converter first, I expect retrofit to try to use it and fall back to GSON if it detects that given type is not recognized. There is a problem with this in your library and I think it can be improved:

One of the legacy API calls that crashes looks like this:

```
public interface ResourceApi {

    @GET("user/{userId}/allowed/{processId}")
    Observable<ApiResponse<Resource<AdPermission>>> fetchAdPermission(@Path("userId"), String userId, @Path("processId") String processId, @Nullable @Query("adId") Long adId);

}
```

API does return JSON-API response, but parsing is done manually via gson, with legacy `Resource` and `ApiResponse` classes. I still want this endpoint to work this way until we find time to migrate everything to your converter.

When `responseBodyConverter` within `JSONAPIConverterFactory` is called, I think you should return `null` to forward the deserialization to GSON, but the app crashes within `RetrofitType` class.

The code enters `initialize(Type type)` method and crashes in line:
`this.type = (Class<?>) typeArgs[0];` with a class cast exception.
I think it happens because I use ParameterizedType within ParameterizedType  and `RetrofitType` class  only works for a single nesting of ParameterizedTypes... Please check this out.

**Comment by jasminb** [comment]
Hey,

I will take a detailed look at this, in the meantime, there is a method to register `alternative` factory to the `JSONAPIConverterFactory`:

```java
jsonApiConverterFactory.setAlternativeFactory(yourGSONFactory);
```

**Comment by bradroid** [comment]
Hi, thanks for proposition but it won't work, because it crashes in the first line of `responseBodyConverter` implementation, when trying to create a RetrofitType instance. Alternative factory is used few lines below, the code will never reach it.

Maybe you should also give your `alternative factory` functionality a second thought, because I think the same behavior can be achieved by setting multiple factories to `Retrofit` instance.

**Comment by bradroid** [comment]
I messed up the order in the first post (it doesn't resolve the problem, only the code example was incorrect, my code still crashes), it should be like this:

```
  retrofit = new Retrofit.Builder()
                .client(client)
                .baseUrl(baseUrl)
                .addConverterFactory(jsonApiConverterFactory)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
```

This way, retrofit will prefer to use jsonApi converter, but fallback to gson if jsonApi converter signals it can not handle the conversion (e.g. returns null from `requestBodyConverter` method). It's the same as alternative factory you offer, already built into retrofit :)

**Comment by bradroid** [comment]
@jasminb Hi, any update for this? Do you accept pull requests... I might find the time to fix this, unless you already started :)

**Comment by jasminb** [comment]
Hey @bradroid, feel free to fix it and create a PR.

**Comment by bradroid** [comment]
@jasminb https://github.com/jasminb/jsonapi-converter/pull/132

It fixes the issue but I think we can design the whole class better to avoid any problems, please check it out.

**Comment by jasminb** [comment]
Hey @bradroid,

Feel free to create another PR that handles your concerns on design issues.

**Comment by jasminb** [comment]
Closing as the issue was fixed with #132
